static/js/narrow.js

```diff
@@ -1053,4 +1033,58 @@ exports.hide_empty_narrow_message = function () {
     $("#left_bar_compose_reply_button_big").removeAttr("disabled");
 };
 
+exports.show_history_limit_notice = function () {
+    $(".top-messages-logo").hide();
+    $(".history-limited-box").show();
+    exports.hide_empty_narrow_message();
+};
+
+exports.hide_history_limit_notice = function () {
+    $(".top-messages-logo").show();
+    $(".history-limited-box").hide();
+};
+
+exports.hide_end_of_results_notice = function () {
+    $(".all-messages-search-caution").hide();
+};
+
+exports.show_end_of_results_notice = function () {
+    $(".all-messages-search-caution").show();
+    // Set the link to point to this search with streams:public added.
+    // It's a bit hacky to use the href, but
+    // !filter.includes_full_stream_history() implies streams:public
+    // wasn't already present.
+    $(".all-messages-search-caution a.search-shared-history").attr(
+        "href", window.location.hash.replace("#narrow/", "#narrow/streams/public/")
+    );
+};
+
+exports.update_top_of_narrow_notices = function (msg_list) {
+    if (msg_list !== current_msg_list) {
+        return;
+    }
+
+    if (msg_list.data.fetch_status.has_found_oldest() &&
+        current_msg_list !== home_msg_list) {
+        const filter = narrow_state.filter();
+        // Potentially display the notice that lets users know
+        // that not all messages were searched.  One could
+        // imagine including `filter.is_search()` in these
+        // conditions, but there's a very legitimate use case
+        // for moderation of searching for all messages sent
+        // by a potential spammer user.
+        if (!filter.contains_only_private_messages() &&
+            !filter.includes_full_stream_history() &&
+            !filter.is_personal_filter()) {
+            exports.show_end_of_results_notice();
+        }
+    }
+
+    if (msg_list.data.fetch_status.history_limited()) {
+        exports.show_history_limit_notice();
+    } else {
+        exports.hide_history_limit_notice();
+    }
+};
+
```

Comment:
I think rather than moving things here, the last commit should move these to a new `static/js/message_list_notices.js`; `narrow.js` is already pretty giant, so we shouldn't be moving things that could be a separate module into it.